File: LC303RangeSumQueryImmutable.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/range-sum-query-immutable/
 */
public class LC303RangeSumQueryImmutable {
    public class NumArray {
        private int[] sums;

        public NumArray(int[] nums) {
            sums = new int[nums.length];
            int sum = 0;
            for (int i = 0; i < nums.length; i++) {
                sum = sums[i] = sum + nums[i];
            }
        }

        public int sumRange(int i, int j) {
            int sum = 0;
            if (0 < i) {
                sum = -sums[i - 1];
            }
            sum += sums[j];
            return sum;
        }
    }

    @Test
    public void test_0__2() throws Exception {
        int[] nums = {-2, 0, 3, -5, 2, -1};
        assertEquals(1, new NumArray(nums).sumRange(0, 2));
    }

    @Test
    public void test_2__5() throws Exception {
        int[] nums = {-2, 0, 3, -5, 2, -1};
        assertEquals(-1, new NumArray(nums).sumRange(2, 5));
    }

    @Test
    public void test_0__5() throws Exception {
        int[] nums = {-2, 0, 3, -5, 2, -1};
        assertEquals(-3, new NumArray(nums).sumRange(0, 5));
    }
}

File: algorithms/code/leetcode/lc303_range_sum_query_immutable/lc303_range_sum_query_immutable.py
#!/usr/bin/env python3
# https://leetcode.com/problems/range-sum-query-immutable/

import unittest
from typing import List


class NumArray:
    def __init__(self, nums: List[int]):
        self.__sums = [0] * len(nums)
        summed = 0
        for i, nums_i in enumerate(nums):
            summed += nums_i
            self.__sums[i] = summed

    def sumRange(self, left: int, right: int) -> int:
        summed = 0
        if left > 0:
            summed = -self.__sums[left - 1]
        summed += self.__sums[right]
        return summed


class TestCode(unittest.TestCase):
    def test_0__2(self):
        nums = [-2, 0, 3, -5, 2, -1]
        self.assertEqual(1, NumArray(nums).sumRange(0, 2))

    def test_2__5(self):
        nums = [-2, 0, 3, -5, 2, -1]
        self.assertEqual(-1, NumArray(nums).sumRange(2, 5))

    def test_0__5(self):
        nums = [-2, 0, 3, -5, 2, -1]
        self.assertEqual(-3, NumArray(nums).sumRange(0, 5))
